modApiServer/test/org/aion/api/server/ApiAionTests.java

```diff
@@ -0,0 +1,398 @@
+
+        @Override
+        protected void onBlock(AionBlockSummary cbs) {
+            onBlockFlag = true;
+        }
+
+        @Override
+        protected void pendingTxReceived(ITransaction _tx) {
+            pendingRcvdFlag = true;
+        }
+
+        @Override
+        protected void pendingTxUpdate(ITxReceipt _txRcpt, EventTx.STATE _state) {
+            pendingUpdateFlag = true;
+        }
+
+        public boolean allFlagsSet() {
+            return (onBlockFlag && pendingRcvdFlag && pendingUpdateFlag);
+        }
+
+
+        public ApiAionImpl(AionImpl impl) {
+            super(impl);
+            onBlockFlag = false;
+            pendingRcvdFlag = false;
+            pendingUpdateFlag = false;
+        }
+
+        public void addEvents() {
+            EventTx pendingRcvd = new EventTx(EventTx.CALLBACK.PENDINGTXRECEIVED0);
+            AionTransaction tx = new AionTransaction(null);
+            List l1 = new ArrayList<ITransaction>();
+            l1.add(tx);
+            l1.add(tx);
+            l1.add(tx);
+            pendingRcvd.setFuncArgs(Collections.singletonList(l1));
+
+            ees.add(pendingRcvd);
+
+            EventTx pendingUpdate = new EventTx(EventTx.CALLBACK.PENDINGTXUPDATE0);
+            List l2 = new ArrayList<>();
+            l2.add(new AionTxReceipt());
+            l2.add(-1);
+            pendingUpdate.setFuncArgs(l2);
+
+            ees.add(pendingUpdate);
+
+            EventBlock evBlock = new EventBlock(EventBlock.CALLBACK.ONBLOCK0);
+            AionBlockSummary abs = new AionBlockSummary(null, null, null, null);
+            evBlock.setFuncArgs(Collections.singletonList(abs));
+
+            ees.add(evBlock);
+
+            //provokes exception in EpApi.run()
+            ees.add(new EventBlock(EventBlock.CALLBACK.ONBLOCK0));
+            ees.add(new EventDummy());
+        }
+
+    }
+
+    @Test
+    public void TestCreate() {
+        System.out.println("run TestCreate.");
+        AionImpl impl = AionImpl.inst();
+        ApiAionTests.ApiAionImpl api = new ApiAionTests.ApiAionImpl(impl);
+        System.out.println("API Version = " + api.getApiVersion());
+        assertNotNull(api.getInstalledFltrs());
+    }
+
+    @Test
+    public void TestInitNrgOracle() {
+        System.out.println("run TestInitNrgOracle.");
+        AionImpl impl = AionImpl.inst();
+        ApiAionTests.ApiAionImpl api = new ApiAionTests.ApiAionImpl(impl);
+        api.initNrgOracle(impl);
+        assertNotNull(api.getNrgOracle());
+        // Initing a second time should not create a new NrgOracle
+        api.initNrgOracle(impl);
+        assertNotNull(api.getNrgOracle());
+    }
+
+    @Test
+    public void TestStartES() {
+        System.out.println("run TestStartES.");
+        AionImpl impl = AionImpl.inst();
+        ApiAionTests.ApiAionImpl api = new ApiAionTests.ApiAionImpl(impl);
+        api.startES("thName");
+        api.addEvents();
+        try
+        {
+            Thread.sleep(5000);
```

Comment:
Can we avoid sleeping here? Or at least for 5 seconds? Also, if the exception is thrown would it cause your assert to fail? If so, a print message or something could be useful otherwise the failure will appear very cryptic

Reply:
It's not clear to me why the sleep is needed.  Could you explain that?

I agree about the InterruptedException... actually, I'd probably just delete that catch and add 'throws Exception' to the method signature

Reply:
so we have a race-condition here...the test only passes if the events dispatched in addEvents() are executed before we pass onto the allFlagsSet assertion...My hacky way of ensuring that always happens is by waiting 5 seconds (which is almost definitely excessive).

what's the "correct" way to enforce this? use a lock of some sort?

ps: sorry i'm weak at concurrency-related stuff :P